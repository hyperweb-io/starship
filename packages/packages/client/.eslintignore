node_modules/
dist/
main/
module/
coverage/
prompt.js
