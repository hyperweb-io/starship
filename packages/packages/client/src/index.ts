export * from './client';
export * from './installer';
