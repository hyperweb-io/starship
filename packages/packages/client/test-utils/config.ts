import { readFileSync } from 'fs';
import * as yaml from 'js-yaml';
import { join, resolve } from 'path';

import { StarshipConfig } from '../src/config';

export const fixtureDir = resolve(join(__dirname, '/../../../__fixtures__'));
export const outputDir = resolve(join(__dirname, '/../../../__output__'));

function loadConfig(filename: string) {
  const configPath = join(fixtureDir, filename);
  const configAsYaml = readFileSync(configPath, 'utf-8');
  const config: StarshipConfig = yaml.load(configAsYaml) as StarshipConfig;
  return {
    configPath,
    configAsYaml,
    config
  };
}

const localConfig = loadConfig('local-config.yaml');
const config = loadConfig('config.yaml');

export { config, localConfig };
