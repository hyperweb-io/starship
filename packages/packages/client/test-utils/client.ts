import { relative } from 'path';
import strip from 'strip-ansi';

import { StarshipClient } from '../src';
import { config } from './config';

interface TestClientCtx {
  commands: string[];
  logs: string[];
  code: number;
}

export const createClient = () => {
  const ctx: TestClientCtx = {
    commands: [],
    logs: [],
    code: -1
  };

  const client = new StarshipClient({
    name: 'osmojs',
    config: relative(process.cwd(), config.configPath)
  });

  const handler = {
    get(target: any, prop: string, _receiver: any) {
      const originalMethod = target[prop];
      if (typeof originalMethod === 'function') {
        return function (...args: any[]) {
          const argsString = args
            .map((arg) => strip(JSON.stringify(arg)))
            .join(', ');
          ctx.logs.push(`Call: ${prop}(${argsString})`);
          // if you want to see nested Call, replace target with this
          // I double checked both this and target, it does not call the exec in the methods when used internally
          return originalMethod.apply(target, args);
        };
      }
      return originalMethod;
    }
  };

  const proxiedClient: StarshipClient = new Proxy(client, handler);

  // Overriding the exit method
  // @ts-expect-error - Ignore lint error
  proxiedClient.exit = (code: number) => {
    ctx.code = code;
  };

  // @ts-expect-error - Ignore lint error
  proxiedClient.ensureFileExists = (_filename: string) => {};

  // Overriding the exec method
  // @ts-expect-error - Ignore lint error
  proxiedClient.exec = (cmd: string[]) => {
    // @ts-expect-error - Ignore lint error
    client.checkDependencies();
    ctx.commands.push(cmd.join(' '));
    ctx.logs.push(cmd.join(' '));
    return '';
  };

  // Overriding the log method
  // @ts-expect-error - Ignore lint error
  proxiedClient.log = (cmd: string) => {
    const str = strip(cmd);
    if (/\n/.test(str)) {
      ctx.logs.push('Log⬇\n' + str + '\nEndLog⬆');
    } else {
      ctx.logs.push('Log: ' + str);
    }
  };

  return {
    client: proxiedClient,
    ctx
  };
};

export const expectClient = (ctx: TestClientCtx, code: number) => {
  expect(ctx.logs.join('\n')).toMatchSnapshot();
  expect(ctx.commands.join('\n')).toMatchSnapshot();
  expect(ctx.code).toBe(code);
};
