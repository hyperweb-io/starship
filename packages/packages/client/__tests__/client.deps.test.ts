import { createClient, expectClient } from '../test-utils/client';

describe('StarshipClient', () => {
  it('missing deps', () => {
    const { client, ctx } = createClient();

    client.dependencies = client.dependencies.map((dep) => {
      if (['kubectl', 'docker'].includes(dep.name)) {
        return {
          ...dep,
          installed: false
        };
      }
      return dep;
    });

    // @ts-expect-error - Ignore lint error
    client.exec(['something']);

    expectClient(ctx, 1);
  });
  it('has all deps', () => {
    const { client, ctx } = createClient();

    client.dependencies.forEach((dep) => (dep.installed = true));

    // @ts-expect-error - Ignore lint error
    client.exec(['something']);

    expectClient(ctx, -1);
  });
});
