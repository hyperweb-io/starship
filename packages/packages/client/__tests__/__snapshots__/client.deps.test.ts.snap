// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StarshipClient has all deps 1`] = `
"Call: exec(["something"])
something"
`;

exports[`StarshipClient has all deps 2`] = `"something"`;

exports[`StarshipClient missing deps 1`] = `
"Call: exec(["something"])
Log⬇
xkubectl
xdocker
✓helm
EndLog⬆
Log⬇

Please install the missing dependencies:
EndLog⬆
Log⬇


kubectl: https://kubernetes.io/docs/tasks/tools/


docker: https://docs.docker.com/get-docker/
For advanced Docker usage and installation on other platforms, see: https://docs.docker.com/get-docker/


EndLog⬆
something"
`;

exports[`StarshipClient missing deps 2`] = `"something"`;
